<?xml version="1.0"?>
<launch>

    <arg name="rover_name" default="scout_1"/>
    <arg name="publish_gt" default="true" />

    <include file="$(find csi_rover_localizer)/launch/visual_odom.launch" />

    <!-- Launch 2D lidar processor for rectification and obstacle detection -->
    <include file="$(find csi_rover_localizer)/launch/lidar_proc.launch" />

    <group ns="localization" >

        <param name="imu_topic" value="/$(arg rover_name)/imu" />
        <param name="rover_name" value="$(arg rover_name)" />
        <param name="dynamic_covariance" value="true" />

        <!-- Public ground truth odom for debugging purposes -->
        <node if="$(arg publish_gt)" pkg="csi_rover_localizer" type="rover_localization_gt.py"
              name="rover_localization_gt" output="screen"/>

        <!-- Odometry node -->
        <!-- Attributions  -->
        <!-- Scaffolding inspired by: https://roverrobotics.com/blogs/guides/fusing-imu-encoders-with-ros-robot-localization -->

        <node pkg="robot_localization" type="ekf_localization_node" name="ekf_odom_node" output="screen" >

            <!-- set up sample rates, etc. -->
            <param name="frequency" value="30"/>
            <param name="sensor_timeout" value="0.1"/>
            <param name="two_d_mode" value="false"/>
            <remap from="odometry/filtered" to="/$(arg rover_name)/ekf_odom"/>
            <param name="base_link_frame" value="$(arg rover_name)_tf/base_footprint"/>
            <param name="publish_tf" value="true" />
            <param name="map_frame" value="map"/>
            <param name="odom_frame" value="odom"/>
            <param name="world_frame" value="odom"/>
            <param name="transform_time_offset" value="0.0"/>
            <param name="gravitational_acceleration" value="1.62" />

            <!-- VISUAL ODOM  -->
            <param name="odom0" value="/rtabmap/visual_odom"/>
            <param name="odom0_differential" value="false" />
            <param name="odom0_relative" value="false" />
            <param name="odom0_queue_size" value="10" />
            <param name="odom0_remove_gravitational_acceleration" value="false" />
            <rosparam param="odom0_config">[true,  true, true,
                                            true, true, true,
                                            true, true, true,
                                            false, false, false,
                                            false, false, false]</rosparam>

            <!-- IMU -->
            <param name="imu0" value="/$(arg rover_name)/imu_rect"/>
            <param name="imu0_differential" value="false" />
            <param name="imu0_relative" value="false" />
            <param name="imu0_queue_size" value="10" />
            <param name="imu0_remove_gravitational_acceleration" value="true" />
            <rosparam param="imu0_config">
                [false,  false, false,
                true , true, false,
                false, false, false,
                false, false, false,
                true, true, true ]
            </rosparam>

            <!-- ENCODER ODOM  -->
            <param name="odom1" value="/$(arg rover_name)/wheel_odom"/>
            <param name="odom1_differential" value="false" />
            <param name="odom1_relative" value="false" />
            <param name="odom1_queue_size" value="10" />
            <param name="odom1_remove_gravitational_acceleration" value="false" />
            <rosparam param="odom1_config">[true, true, true,
                                           false, false, true,
                                           false, false, false,
                                           false, false, false,
                                           false, false, false]</rosparam>

            <!-- LASER ODOM
            <param name="odom0" value="/odom_rf2"/>
            <param name="odom0_differential" value="false" />
            <param name="odom0_relative" value="true" />
            <param name="odom0_queue_size" value="10" />
            <param name="odom0_remove_gravitational_acceleration" value="false" />
            <rosparam param="odom0_config">[true,  true, false,
                                           false, false, true,
                                           false, false, false,
                                           false, false, false,
                                           false, false, false]</rosparam> -->

            <param name="print_diagnostics" value="true" />
            <param name="debug" value="true" />
            <param name="debug_out_file" value="debug_odom_ekf.txt" />
            <rosparam param="process_noise_covariance">
                [0.05, 0,    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0,
                0, 0.05, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,
                0, 0, 0.06, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,
                0, 0, 0, 0.03, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0.03, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0.06, 0, 0, 0,  0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0.025,0, 0,  0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0.025,0,  0, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0.04, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0,  0.01, 0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0.01, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0.02, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0.01, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0.01, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0,  0, 0, 0, 0, 0, 0.005]</rosparam>

            <rosparam param="initial_estimate_covariance">
                [1e-9, 0,    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0,
                0, 1e-9, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,
                0, 0, 1e-9, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 1e-9, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 1e-9, 0, 0, 0, 0, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 1e-9, 0, 0, 0, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 1 , 0, 0, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 1e-9, 0,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 1e-9,    0, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    1e-9, 0, 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 1e-9 , 0, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 1e-9, 0, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 1e-9, 0,
                0, 0, 0, 0, 0, 0, 0, 0, 0, 0,    0, 0, 0, 0, 1e-9]</rosparam>

        </node>

    </group>

</launch>
